orderapi/src/main/java/fi/hel/verkkokauppa/order/listeners/DeadLetterQueueListener.java

```diff
@@ -0,0 +1,137 @@
+
+    /*
+     * This listener is only used in local development,
+     * because the default DLQ in ActiveMQ is named "ActiveMQ.DLQ" instead of "DLQ".
+     */
+    @Profile("local")
+    @JmsListener(destination = "${queue.dlq:ActiveMQ.DLQ}")
+    public void consumeMessageLocal(TextMessage textMessage) throws Exception {
+        try {
+            PaymentMessage message = getPaymentMessageFromTextMessage(textMessage);
+
+            logMessageData(message);
+
+            if (message.getEventType().equals(EventType.PAYMENT_PAID)) {
+                paymentFailedAction(message);
+            }
+        } catch(JsonProcessingException jsonProcessingException) {
+            log.debug(jsonProcessingException.getMessage());
+            log.info("Failed to convert queue message to PaymentMessage type.");
+        }
+    }
+
+    private void paymentFailedAction(PaymentMessage message) {
+        PaymentFailedMessage paymentFailedMessage = PaymentFailedMessage.builder()
+                .eventType(EventType.PAYMENT_FAILED)
+                .paymentId(message.getPaymentId())
+                .eventTimestamp(LocalDateTime.now().toString())
+                .namespace(message.getNamespace())
+                .build();
+
+        sendNotificationToEmail(paymentFailedMessage);
+        sendNotificationService.sendToQueue(paymentFailedMessage, queueConfigurations.getPaymentFailedToProcessQueue());
+    }
+
+    private void sendNotificationToEmail(PaymentFailedMessage paymentFailedMessage) {
+        JSONObject msgJson = new JSONObject();
+        msgJson.put("id", paymentFailedMessage.getPaymentId());
+        msgJson.put("receiver", paymentFailedNotificationEmail);
+        msgJson.put("header", "DLQ queue alert - " + EventType.PAYMENT_FAILED);
+        try {
+            String html = Files.readString(Paths.get(ClassLoader.getSystemResource(EMAIL_TEMPLATE_PATH).toURI()));
+            html = html.replace("#EVENT_TYPE#", paymentFailedMessage.getEventType());
+            html = html.replace("#PAYMENT_ID#", paymentFailedMessage.getPaymentId());
+            html = html.replace("#NAMESPACE#", paymentFailedMessage.getNamespace());
+            html = html.replace("#EVENT_PAYLOAD#", mapper.writeValueAsString(paymentFailedMessage));
+
+            msgJson.put("body", html);
+        } catch (IOException | URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+
+        log.info("Payment with id {} failed. Sending email notification to {}", paymentFailedMessage.getPaymentId(), paymentFailedNotificationEmail);
+        restServiceClient.makePostCall(messageServiceUrl + "send/email", msgJson.toString());
+    }
+
+    private PaymentMessage getPaymentMessageFromTextMessage(TextMessage textMessage) throws JMSException, JsonProcessingException {
+        String jsonMessage = textMessage.getText();
+        log.info(DeadLetterQueueListener.class + "{}", jsonMessage);
+
+        return mapper.readValue(jsonMessage, PaymentMessage.class);
```

Comment:
What happens if some else Message.class happens to end to DLQ queue? That causes those exceptions but should we just ignore and log them in my opinion.

Reply:
The message error thrown here is caught at the methods that call this method (**consumeMessage** and **consumeMessageLocal**).

Reply:
I added the payment notification a property of MsgType = PAYMENT_PAID so that we can leverage the JMS selector to only consume PAYMENT_PAID messages from DLQ. This way we dont have to implement separate logic to filter the right messages in DLQ listener. Property is added in message convertion and send phase in **SendNotificationService.sendPaymentMessageNotification(...)** line 71 (_fi.hel.verkkokauppa.common.queue.service.SendNotificationService#sendPaymentMessageNotification_).